java/impl/src/main/java/net/sf/mmm/code/impl/java/source/maven/JavaSourceProviderUsingMaven.java

```diff
@@ -129,71 +140,276 @@ private BaseSource createSource(Dependency dependency, File byteCodeArtifact) {
+      dependenciesURLs = getDependenciesURLS(source.getModel(), dependenciesURLs, 0);
+
+      try {
+        URL[] dependencies = new URL[dependenciesURLs.size()];
+        dependencies = dependenciesURLs.toArray(dependencies);
+
+        MavenClassLoader mvnClassLoader = new MavenClassLoader(Thread.currentThread().getContextClassLoader(),
+            dependencies);
+
+        return new JavaExtendedContext(source, provider, mvnClassLoader);
+      } catch (Exception e) {
+        throw new IllegalStateException(e);
+      }
+    } else {
+      return new JavaExtendedContext(source, provider, null);
+    }
+  }
+
+  /**
+   * Retrieves the URL where the eclipse-target folder should be located. It contains bytecode. This is useful for
+   * devon4j projects, but will not harm other projects.
+   * 
+   * @param byteCodeLocation current location of the byte code
+   * @return the URL with the location of eclipse-target folder
+   * @throws MalformedURLException throws {@link MalformedURLException}
+   */
+  private URL getEclipseByteCodeLocation(File byteCodeLocation) throws MalformedURLException {
+
+    String[] segments = getPathSegmentsFromFile(byteCodeLocation);
+
+    if (segments.length >= 3) {
+      segments[segments.length - 2] = "eclipse-target";
+    }
+
+    String eclipseTargetLocation = "";
+    for (String segment : segments) {
+      eclipseTargetLocation = eclipseTargetLocation + segment + "/";
+    }
+    return new URL(eclipseTargetLocation);
+  }
+
+  /**
+   * Tries to find the parent POM of this project in order to retrieve the modules defined-
+   * 
+   * @param location current project, in which we want to check if it has a parent POM with modules
+   * @param dependenciesURLs list of previous dependencies URLs
+   * @return list of dependencies URLs retrieved from the modules and the parent POM
+   */
+  private ArrayList<URL> getModulesDependenciesURLS(File location, ArrayList<URL> dependenciesURLs) {
+
+    Model model = parseModel(location);
+
+    if (model == null) {
+      return dependenciesURLs;
+    }
+
+    try {
+      while (model.getParent() != null) {
+        File parentPom = location.toPath().resolve(model.getParent().getRelativePath()).toFile().getCanonicalFile();
+        Model parentModel = parseModel(parentPom);
+
+        if (parentModel == null) {
+          return dependenciesURLs;
+        }
+
+        for (String module : parentModel.getModules()) {
+          String[] segments = getPathSegmentsFromFile(parentPom);
+          if (segments.length - 1 == 0) {
+            break;
+          }
+
+          segments[segments.length - 1] = module;
+
+          String moduleLocation = "";
+          int i = 0; // I need this to get the file not as URI
+          for (String segment : segments) {
+            if (i != 0) {
+              moduleLocation = moduleLocation + segment + "/";
+            }
+            i++;
+          }
```

Comment:
do not fiddle with path via such string operations. Instead simply use Java File or Path API.